lib/uplink/encryption.go

```diff
@@ -4,20 +4,184 @@
+// in the new context is just enough to allow someone to access all of the given
+// restrictions but no more.
+func (s *EncryptionCtx) Restrict(apiKey APIKey, restrictions ...EncryptionRestriction) (APIKey, *EncryptionCtx, error) {
+	if len(restrictions) == 0 {
+		// Should the function signature be
+		// func (s *EncryptionCtx) Restrict(apiKey APIKey, restriction EncryptionRestriction, restrictions ...EncryptionRestriction) (APIKey, *EncryptionCtx, error) {
+		// so we don't have to do this test?
+		return APIKey{}, nil, errs.New("at least one restriction required")
+	}
+
+	caveat := macaroon.Caveat{}
+	encCtx := NewEncryptionCtx()
+
+	for _, res := range restrictions {
+		unencPath := paths.NewUnencrypted(res.PathPrefix)
+		cipher := storj.AESGCM // TODO(jeff): pick the right path cipher
+
+		encPath, err := encryption.StoreEncryptPath(res.Bucket, unencPath, cipher, s.store)
+		if err != nil {
+			return APIKey{}, nil, err
+		}
+		derivedKey, err := encryption.StoreDerivePathKey(res.Bucket, unencPath, s.store)
+		if err != nil {
+			return APIKey{}, nil, err
+		}
+
+		if err := encCtx.store.Add(res.Bucket, unencPath, encPath, *derivedKey); err != nil {
+			return APIKey{}, nil, err
+		}
+		caveat.AllowedPaths = append(caveat.AllowedPaths, &macaroon.Caveat_Path{
+			Bucket:              []byte(res.Bucket),
+			EncryptedPathPrefix: []byte(encPath.Raw()),
+		})
+	}
+
+	apiKey, err := apiKey.Restrict(caveat)
+	if err != nil {
+		return APIKey{}, nil, err
+	}
+
+	return apiKey, encCtx, nil
+}
+
+// Serialize turns an EncryptionCtx into base58
+func (s *EncryptionCtx) Serialize() (string, error) {
+	var storeEntries []*pb.EncryptionCtx_StoreEntry
+	err := s.store.Iterate(func(bucket string, unenc paths.Unencrypted, enc paths.Encrypted, key storj.Key) error {
+		storeEntries = append(storeEntries, &pb.EncryptionCtx_StoreEntry{
+			Bucket:          []byte(bucket),
+			UnencryptedPath: []byte(unenc.Raw()),
+			EncryptedPath:   []byte(enc.Raw()),
+			Key:             key[:],
+		})
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	var defaultKey []byte
+	if key := s.store.GetDefaultKey(); key != nil {
+		defaultKey = key[:]
+	}
+
+	data, err := proto.Marshal(&pb.EncryptionCtx{
+		DefaultKey:   defaultKey,
+		StoreEntries: storeEntries,
+	})
+	if err != nil {
+		return "", errs.New("unable to marshal encryption ctx: %v", err)
+	}
+
+	return base58.CheckEncode(data, 0), nil
+
+}
+
+// ParseEncryptionCtx parses a base58 serialized encryption context into a working one.
+func ParseEncryptionCtx(b58data string) (*EncryptionCtx, error) {
+	data, version, err := base58.CheckDecode(b58data)
+	if err != nil || version != 0 {
+		return nil, errs.New("invalid encryption context format")
+	}
+
+	p := new(pb.EncryptionCtx)
+	if err := proto.Unmarshal(data, p); err != nil {
+		return nil, errs.New("unable to unmarshal encryption context: %v", err)
+	}
+
+	encCtx := NewEncryptionCtx()
+
+	if len(p.DefaultKey) > 0 {
```

Comment:
nit: seems like we should ensure that protobuf already unserializes into the right type so we don't have to check this. e.g. There's NodeID that already does this.